### PR TITLE
ASoC: SOF: Intel: hda: call codec wake at chip init

### PR DESCRIPTION
**Pull request by kv2019i**
Further align HDA init sequence to the legacy non-DSP HDA driver by
calling snd_hdac_set_codec_wakeup() during the chip init sequence.

One part of fix needed for #1847 (other part is a i915 driver change)

### CODE REVIEW

File: sound/soc/sof/intel/hda-ctrl.c [hunk 1]
```diff
@@ -219,7 +223,8 @@ int hda_dsp_ctrl_init_chip(struct snd_sof_dev *sdev, bool full_reset)
 	/* check to see if controller is ready */
 	if (!snd_hdac_chip_readb(bus, GCTL)) {
 		dev_dbg(bus->dev, "controller not ready!\n");
-		return -EBUSY;
+		ret = -EBUSY;
+		goto err;
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-ctrl.c:227]: this changes the sequence now, in the err: handling you will do a 	hda_dsp_ctrl_misc_clock_gating(sdev, true); which wasn't done before. Or the err: label location is off by one line?

I would suggest splitting the error flow corrections from the chip_init stuff, that's confusing here, no?

File: sound/soc/sof/intel/hda-ctrl.c [hunk 2]
```diff
@@ -285,7 +290,11 @@ int hda_dsp_ctrl_init_chip(struct snd_sof_dev *sdev, bool full_reset)
 
 	bus->chip_init = true;
 
+err:
 	hda_dsp_ctrl_misc_clock_gating(sdev, true);
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-ctrl.c:295]: should err: be line 295?

File: sound/soc/sof/intel/hda-ctrl.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-ctrl.c:227]: @plbossart I think it was a miss in the current code to not set clock_gating to true in case of error no?

Comment (kv2019i) [reply sound/soc/sof/intel/hda-ctrl.c:227]: @ranj063 ack indeed. The existing code leaves the misc clock gating off if init fails. When I added the error case for the codec wakeline, it looked a bit wrong to unroll only part of the change. But ack, let me check the impacts again and submit as separate patch to make this more clear. The codec wakeline we have to unroll for sure as that has impact to other IP. The misc clock control is less obvious as audio IP is left anyways into uninitialized state when this function fails.

File: sound/soc/sof/intel/hda-ctrl.c (hunk 2 above)
Comment (kv2019i) [reply sound/soc/sof/intel/hda-ctrl.c:295]: @plbossart so this was indeed on purpose here, but let me rework this to make this clearer

File: sound/soc/sof/intel/hda-ctrl.c (hunk 1 above)
Comment (kv2019i) [reply sound/soc/sof/intel/hda-ctrl.c:227]: @plbossart @ranj063 now split into a separate patch

Comment (plbossart) [reply sound/soc/sof/intel/hda-ctrl.c:227]: ok makes sense then!

File: sound/soc/sof/intel/hda-ctrl.c [hunk 3]
```diff
@@ -288,6 +292,9 @@ int hda_dsp_ctrl_init_chip(struct snd_sof_dev *sdev, bool full_reset)
 
 err:
 	hda_dsp_ctrl_misc_clock_gating(sdev, true);
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+	snd_hdac_set_codec_wakeup(bus, false);
+#endif
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-ctrl.c:297]: @kv2019i does the order between the clock gating and codec wakeup matter?

I was wondering if for extra robustness we'd want the clock gating off, with the IP having no inhibitions before detecting a wakeup?

Comment (kv2019i) [reply sound/soc/sof/intel/hda-ctrl.c:297]: @plbossart No documented dependency, but this is the order in hda_intel.c so I followed that to the letter w.r.t rest of chip init actions.

Comment (plbossart) [reply sound/soc/sof/intel/hda-ctrl.c:297]: ok, that's fair, and that's what your commit message says, so we're good.

**Pull request merged**